Translate api/index - options/lifecycle hooks (#63)
1950Labs/vuejs.org: src/v2/api/index.md

  - **Vea También:** [Función Render](../guide/render-function.html)

## Opciones / Hooks de Ciclo de Vida

> <p class="tip">Todos los _hooks_ de ciclo de vida tienen automáticamente su contexto `this` vinculado a la instancia, de modo que usted puede acceder a datos, propiedades computadas y métodos. Esto significa que __no debe usar una _arrow function_ para definir un método de ciclo de vida__ (por ejemplo, `created: () => this.fetchTodos ()`). La razón es que las _arrow functions_ vinculan el contexto principal, por lo que `this` no será la instancia de Vue como espera y `this.fetchTodos` no estará definido.</p>

Reviewer: I checked at mdn [arrow functions](https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Funciones/Arrow_functions) in the spanish version and they call them: `Funciones Flecha`. I'm good if we want to keep it like that but simply wanted to share this.

In case you prefer changing it, quick action:
```suggestion
<p class="tip">Todos los _hooks_ de ciclo de vida tienen automáticamente su contexto `this` vinculado a la instancia, de modo que usted puede acceder a datos, propiedades computadas y métodos. Esto significa que __no debe usar una _función flecha_ para definir un método de ciclo de vida__ (por ejemplo, `created: () => this.fetchTodos ()`). La razón es que las _funciones flecha_ vinculan el contexto principal, por lo que `this` no será la instancia de Vue como espera y `this.fetchTodos` no estará definido.</p>
```

Author: We are using `arrow functions` in several places.
We should have a list of doubts or improvements to discuss with the team. 
For now we can keep it as is if you agree.

### beforeCreate

- **Tipo:** `Function`

- **Detalles:**

  Es invocado sincrónicamente, inmediatamente después de que se ha inicializado la instancia, antes de la observación de datos y la configuración de eventos y observadores.

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### created

- **Tipo:** `Function`

- **Detalles:**

>   Se invoca sincrónicamente después de crear la instancia. En este punto, la instancia ha terminado las opciones de procesamiento, lo que significa que se ha configurado: observación de datos, propiedades computadas, métodos, _callbacks_ de observadores y eventos. Sin embargo, la fase de montaje no ha comenzado y la propiedad `$el` aún no estará disponible.

Reviewer 2: I think that "At this state, the instance has finished processing the options" is trying to say that the getters and setters for the instance have been created (for the reactivity system). Having this in mind, I think that translating it to "En este punto, la instancia ha terminado las opciones de procesamiento" looks a bit odd. What do you think about this suggestion:
```suggestion
  Se invoca sincrónicamente después de crear la instancia. En este punto, la instancia ha terminado de procesar las opciones, lo que significa que se ha configurado: observación de datos, propiedades computadas, métodos, _callbacks_ de observadores y eventos. Sin embargo, la fase de montaje no ha comenzado y la propiedad `$el` aún no estará disponible.
```

Author: yes, looks better

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### beforeMount

- **Tipo:** `Function`

- **Detalles:**

>   Invocado justo antes de que comience el montaje: la función `render` está a punto de llamarse por primera vez.

Reviewer 2: "llamarse por primera vez" or "ser llamada por primera vez"? 🤔 Not sure which one is better, your call on this 😄

Reviewer: I vote for: `ser llamada por primera vez` but I'm good with both

Author: yes.. I like `ser llamada por primera vez`

  **Este _hook_ no se invoca durante la renderización del lado del servidor.**

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### mounted

- **Tipo:** `Function`

- **Detalles:**

  Es invocado después de que se ha montado la instancia, donde `el` se reemplaza por el recién creado `vm.$el`. Si la instancia raíz está montada en un elemento en el documento, `vm.$el` también estará en el documento cuando se llame a `mounted`.

  Tenga en cuenta que `mounted` **no** garantiza que todos los componentes secundarios también se hayan montado. Si desea esperar hasta que se haya procesado toda la vista, puede usar [vm.$nextTick](#vm-nextTick) dentro de `mounted`:

  ``` js
  mounted: function () {
    this.$nextTick(function () {
      // Código que se ejecutará solo después de

>       // haber procesado la vista completa

Reviewer 2: What do you think about changing this to the following:
```suggestion
      // haber renderizado la vista completa
```

I suggest this in order to be consistent with the comment on line 839

Author: yes, good idea

    })
  }
  ```

  **Este _hook_ no se invoca durante la renderización del lado del servidor.**

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### beforeUpdate

- **Tipo:** `Function`

- **Detalles:**

  Es invocado cuando los datos cambian, antes de que se actualize el DOM. Este es un buen lugar para acceder al DOM existente antes de una actualización, por ejemplo para eliminar _listeners_ de eventos agregados manualmente.

  **Este _hook_ no se invoca durante la renderización del lado del servidor, porque solo la renderización inicial se realiza del lado del servidor.**

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### updated

- **Tipo:** `Function`

- **Detalles:**

>   Invocado después de un cambio de datos hace que el DOM virtual se vuelva a procesar y actualizar.

Reviewer 2: What do you think about this suggestion?
```suggestion
  Invocado después de que un cambio de datos hace que el DOM virtual se vuelva a procesar y actualizar.
```

>   El DOM del componente se habrá actualizado cuando se llame a este _hook_, por lo que puede realizar operaciones dependientes del DOM aquí. Sin embargo, en la mayoría de los casos, debe evitar cambiar el estado dentro del _hook_. Para reaccionar a los cambios de estado, generalmente es mejor usar una [propiedad computada](#computed) o [watcher](#watch) en su lugar.

Reviewer 2: Missing word 😅 
```suggestion
  El DOM del componente se habrá actualizado cuando se llame a este _hook_, por lo que puede realizar operaciones dependientes del DOM aquí. Sin embargo, en la mayoría de los casos, debe evitar cambiar el estado dentro del _hook_. Para reaccionar a los cambios de estado, generalmente es mejor usar una [propiedad computada](#computed) o un [watcher](#watch) en su lugar.
```

- **Vea también:**
  - [Componentes Built-in - keep-alive](#keep-alive)
  - [Componentes Dinámicos - keep-alive](../guide/components.html#keep-alive)

### beforeDestroy

- **Tipo:** `Function`

- **Detalles:**

  Invocado justo antes de que se destruya una instancia de Vue. En esta etapa, la instancia todavía es completamente funcional.

  **Este _hook_ no se invoca durante la renderización del lado del servidor.**

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### destroyed

- **Tipo:** `Function`

- **Detalles:**

  Invocado después de que una instancia de Vue ha sido destruida. Cuando se llama a este _hook_, todas las directivas de la instancia de Vue se han desvinculado, todos los _listeners_ de eventos se han eliminado y todas las instancias de Vue secundarias también se han destruido.

  **Este _hook_ no se invoca durante la renderización del lado del servidor.**

- **Vea también:** [Diagrama de Ciclo de Vida](../guide/instance.html#Lifecycle-Diagram)

### errorCaptured

> Nuevo en 2.5.0+

- **Tipo:** `(err: Error, vm: Component, info: string) => ?boolean`

- **Detalles:**

>   Se invoca cuando se captura un error de cualquier componente descendiente. El _hook_ recibe tres argumentos: el error, la instancia del componente que activó el error y una cadena que contiene información sobre dónde se capturó el error. El _hook_ puede devolver `falso` para evitar que el error se propague aún más.

Reviewer 2: What do you think about translating "triggered" to "disparó" or "lanzó" instead of "activó"?

Author: I think `disparó` works good

  <p class="tip">Puede modificar el estado del componente en este _hook_. Sin embargo, es importante tener condicionales en su _template_ o función de renderización que alterne otro contenido cuando se haya capturado un error; de lo contrario, el componente caerá en un bucle de renderización infinito.</p>

  **Reglas de Propagación de Errores**

  - De forma predeterminada, todos los errores aún se envían al `config.errorHandler` global si está definido, por lo que estos errores aún se pueden informar a un servicio de análisis en un solo lugar.

>   - Si existen varios _hooks_ `errorCaptured` en una cadena de herencia de un componente o en la cadena principal, se invocarán todos en el mismo error.

Reviewer 2: What do you think about specifying that "se invocarán todos en el mismo error" refers to the multiple `errorCaptured` hooks?

Author: Yes, it's about multiple errorCaptured hooks right. 
We can change it to something like: 

`Si existen múltiples _hooks_ errorCaptured en una cadena de herencia de componentes, todos serán invocados por el mismo error.` What do you think? @FLarra @Tintef

Reviewer 2: @sxcamacho That looks good 😄
